### PR TITLE
Loop3d units

### PR DESCRIPTION
**Pull request by nicholascar**
Units required by the Loop3D project.

These units, plus a few more, will appear in a Community Profile about to be PRed across also.

### CODE REVIEW

**Review by steveraysteveray — CHANGES_REQUESTED** [review]
I now realize I need to update our Unit Submission Guide to reflect the fact that we are now including the property qudt:hasDimensionVector in all our Unit instances. Would you mind adding the appropriate values for your new units? (Easily found, of course, via the hasquantitykind/hasDimensionVector path).

Thanks.

**Comment by nicholascar** [comment]
> Would you mind adding the appropriate [qudt:hasDimensionVector] values for your new units?

OK, I've done the best I can here for the moment without perhaps some more examples. 

I used the documentation at [1] of course but it occurs to me that some small updates are needed there (things I couldn't find) but I can't commit to it (since it's a wiki page, not in the repo directly). Can you please provide the general description of the vector before the example, so:

**1. alter the order of listed dimensions to match the vector?**

* Length - meter (m)
* Time - second (s)
* Amount of substance - mole (mole)
* Electric current - ampere (A)
* Temperature - kelvin (K)
* Luminous intensity - candela (cd)
* Mass - kilogram (kg)

 becomes

* Amount of substance - mole (mole)
* Electric current - ampere (A)
* Length - meter (m)
* Luminous intensity - candela (cd)
* Mass - kilogram (kg)
* Temperature - kelvin (K)
* Time - second (s)

**2. Add in this text:**

In QUDT every dimension vector is comprised of each of the above base dimensions and an associated exponent. This way they can be compared easily. An example for the QuantityKindDimensionVector associated with Force is:

becomes

In QUDT every dimension vector is comprised of each of the above base dimensions and an associated exponent in the following way: 
```
(Dimension1 * dimensional_exponent1) * (Dimension2 * dimensional_exponent2) * ...
```
This is repeated for all 7 dimensions and the dimensionless indicator in the following order:

```
Amount of substance, Electric current, Length, Luminous intensity, Mass, Temperature, Time, Dimensionless
```

The dimensions and the dimensionless value are indicated with single letter codes which, for the above dimensions and dimensionless are, in order, as above:

```
A E L I M H T D
```

An example for the QuantityKindDimensionVector associated with Force is:


Finally, I don't think I've nailed the QuantityKind here [3].


[1] https://github.com/qudt/qudt-public-repo/wiki/User-Guide-for-QUDT#3-introducing-dimension-vectors-for-si-units
[2] https://github.com/qudt/qudt-public-repo/wiki/User-Guide-for-QUDT#the-si-dimensions-and-units
[3] https://github.com/qudt/qudt-public-repo/pull/320/files#diff-2149df826d5aade942798543aa01f733022e6040d90c37e60125d55cf5df54c6R8788

**Comment by steveraysteveray** [comment]
@nicholascar, thanks for making it super-easy for me to put your changes into the wiki! That's done.

I will get back to seeing what you did in [3] above as soon as I can.

**Comment by dr-shorthair** [comment]
See #272 for ongoing discussion of relative importance of Unit, Quantity-Kind and DImension-Vector

**Review by steveraysteveray — CHANGES_REQUESTED** [review]
I'm struggling with the idea that unit:SUSCEPTIBILITY_ELEC and unit:SUSCEPTIBILITY_MAG are units rather than QuantityKinds. We already have them defined as quantitykind:ElectricSusceptibility and quantitykind:MagneticSusceptability.

I don't think these two should be added as units.

**Comment by dr-shorthair** [comment]
Elec and mag sus are dimensionless, but observations produce numerical values. Those values have a 'unit of measure' which is conventionally known as the Electrical Susceptibility unit, and the Magnetic Susceptibility unit, respectively. There could be a scale factor (though the conventional one does not). My suggested change would only be to the label.

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 1]
```diff
@@ -18513,6 +18528,30 @@ unit:SurinamDollar
...
+  a qudt:DerivedUnit ;
+  a qudt:Unit ;
+  qudt:plainTextDescription "Electric susceptibility is a dimensionless proportionality constant that indicates the degree of polarization of a dielectric material in response to an applied electric field. Here P = epsilon_0 * chi_e * E. Where epsilon_0 is the electric permittivity of free space (electric constant), P is the polarization density of the material chi_e is the electric susceptibility and E is the electric field." ;
+  qudt:conversionMultiplier 1.0e0 ;
+  qudt:conversionOffset "0.0"^^xsd:double ;
+  qudt:latexDefinition "\\chi_{\\text{e}} = \\frac{{\\mathbf P}}{\\varepsilon_0{\\mathbf E}}"^^qudt:LatexString ;
+  qudt:hasDimensionVector qkdv:A0E0L0I0M0H0T0D1 ;
+  qudt:hasQuantityKind quantitykind:Dimensionless ;
+  rdfs:isDefinedBy <http://qudt.org/2.1/vocab/unit> ;
+  rdfs:label "Electric Susceptibility" ;
```
Comment (dr-shorthair) [inline vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:18541]: Suggest 
```
  rdfs:label "Electric Susceptibility Unit" ;
```

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 2]
```diff
@@ -18513,6 +18528,30 @@ unit:SurinamDollar
   rdfs:isDefinedBy <http://qudt.org/2.1/vocab/unit> ;
   rdfs:label "Surinam Dollar" ;
 .
+unit:SUSCEPTIBILITY_ELEC
+  a qudt:DerivedUnit ;
+  a qudt:Unit ;
+  qudt:plainTextDescription "Electric susceptibility is a dimensionless proportionality constant that indicates the degree of polarization of a dielectric material in response to an applied electric field. Here P = epsilon_0 * chi_e * E. Where epsilon_0 is the electric permittivity of free space (electric constant), P is the polarization density of the material chi_e is the electric susceptibility and E is the electric field." ;
+  qudt:conversionMultiplier 1.0e0 ;
```
Comment (dr-shorthair) [inline vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:18535]: Suggest using decimal
```
  qudt:conversionMultiplier 1.0 ;
```

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 3]
```diff
@@ -18513,6 +18528,30 @@ unit:SurinamDollar
...
+  a qudt:DerivedUnit ;
+  a qudt:Unit ;
+  qudt:plainTextDescription "Magnetic susceptibility is a dimensionless proportionality constant that indicates the degree of magnetization of a material in response to an applied magnetic field. Here M = chi * H. Where M is the magnetization of the material (the magnetic dipole moment per unit volume), measured in amperes per meter, and H is the magnetic field strength, also measured in amperes per meter. Chi is therefore a dimensionless quantity." ;
+  qudt:conversionMultiplier 1.0e0 ;
+  qudt:conversionOffset "0.0"^^xsd:double ;
+  qudt:latexDefinition "\\chi_\\text{v} = \\frac{\\mathbf{M}}{\\mathbf{H}}"^^qudt:LatexString ; 
+  qudt:hasDimensionVector qkdv:A0E0L0I0M0H0T0D1 ;
+  qudt:hasQuantityKind quantitykind:Dimensionless ;
+  rdfs:isDefinedBy <http://qudt.org/2.1/vocab/unit> ;
+  rdfs:label "Magnetic Susceptibility" ;
```
Comment (dr-shorthair) [inline vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:18553]: Suggest 
```
  rdfs:label "Magnetic Susceptibility Unit" ;
```

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 4]
```diff
@@ -18513,6 +18528,30 @@ unit:SurinamDollar
...
+  qudt:hasDimensionVector qkdv:A0E0L0I0M0H0T0D1 ;
+  qudt:hasQuantityKind quantitykind:Dimensionless ;
+  rdfs:isDefinedBy <http://qudt.org/2.1/vocab/unit> ;
+  rdfs:label "Electric Susceptibility" ;
+.
+unit:SUSCEPTIBILITY_MAG
+  a qudt:DerivedUnit ;
+  a qudt:Unit ;
+  qudt:plainTextDescription "Magnetic susceptibility is a dimensionless proportionality constant that indicates the degree of magnetization of a material in response to an applied magnetic field. Here M = chi * H. Where M is the magnetization of the material (the magnetic dipole moment per unit volume), measured in amperes per meter, and H is the magnetic field strength, also measured in amperes per meter. Chi is therefore a dimensionless quantity." ;
+  qudt:conversionMultiplier 1.0e0 ;
```
Comment (dr-shorthair) [inline vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:18547]: Suggest using decimal
```
  qudt:conversionMultiplier 1.0 ;
```

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 5]
```diff
@@ -18513,6 +18528,30 @@ unit:SurinamDollar
...
+  qudt:conversionOffset "0.0"^^xsd:double ;
```
Comment (dr-shorthair) [inline vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:18548]: Suggest using decimal
```
  qudt:conversionOffset 1.0 ;
```

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 6]
```diff
@@ -18513,6 +18528,30 @@ unit:SurinamDollar
   rdfs:isDefinedBy <http://qudt.org/2.1/vocab/unit> ;
   rdfs:label "Surinam Dollar" ;
 .
+unit:SUSCEPTIBILITY_ELEC
+  a qudt:DerivedUnit ;
+  a qudt:Unit ;
+  qudt:plainTextDescription "Electric susceptibility is a dimensionless proportionality constant that indicates the degree of polarization of a dielectric material in response to an applied electric field. Here P = epsilon_0 * chi_e * E. Where epsilon_0 is the electric permittivity of free space (electric constant), P is the polarization density of the material chi_e is the electric susceptibility and E is the electric field." ;
+  qudt:conversionMultiplier 1.0e0 ;
+  qudt:conversionOffset "0.0"^^xsd:double ;
```
Comment (dr-shorthair) [inline vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:18536]: Suggest using decimal
```
  qudt:conversionOffset 1.0 ;
```

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 7]
```diff
@@ -8777,6 +8777,21 @@ unit:KiloGM-M2
   rdfs:isDefinedBy <http://qudt.org/2.1/vocab/unit> ;
   rdfs:label "Kilogram Meter Squared" ;
 .
+unit:KiloGM-PER-M3
+  a qudt:DerivedUnit ;
+  a qudt:Unit ;
+  qudt:plainTextDescription "Kilogram per cubic metre is an SI derived unit of density, defined by mass in kilograms divided by volume in cubic metres. The official SI symbolic abbreviation is \\(kg \\cdot m^{-3}\\), or equivalently either \\(kg/m^3\\)."^^qudt:LatexString ;
+  qudt:conversionMultiplier 1.0e0 ;
+  qudt:conversionOffset "0.0"^^xsd:double ;
```
Comment (dr-shorthair) [inline vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:8785]: Suggest using decimal
```
  qudt:conversionOffset 1.0 ;
```

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 8]
```diff
@@ -8777,6 +8777,21 @@ unit:KiloGM-M2
   rdfs:isDefinedBy <http://qudt.org/2.1/vocab/unit> ;
   rdfs:label "Kilogram Meter Squared" ;
 .
+unit:KiloGM-PER-M3
+  a qudt:DerivedUnit ;
+  a qudt:Unit ;
+  qudt:plainTextDescription "Kilogram per cubic metre is an SI derived unit of density, defined by mass in kilograms divided by volume in cubic metres. The official SI symbolic abbreviation is \\(kg \\cdot m^{-3}\\), or equivalently either \\(kg/m^3\\)."^^qudt:LatexString ;
+  qudt:conversionMultiplier 1.0e0 ;
```
Comment (dr-shorthair) [inline vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:8784]: Suggest using decimal
```
  qudt:conversionMultiplier 1.0 ;
```

**Comment by dr-shorthair** [comment]
Better to express all multipliers and offsets as `xsd:decimal` - see https://github.com/qudt/qudt-public-repo/issues/307

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl (hunk 3 above)
Comment (nicholascar) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl]: Addressed by Comment https://github.com/qudt/qudt-public-repo/pull/320/commits/78ef2c191aee473ba38b69864919bb5c448a17be

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl (hunk 1 above)
Comment (nicholascar) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl]: Addressed by Comment https://github.com/qudt/qudt-public-repo/pull/320/commits/78ef2c191aee473ba38b69864919bb5c448a17be

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl (hunk 5 above)
Comment (nicholascar) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl]: Addressed by Comment https://github.com/qudt/qudt-public-repo/pull/320/commits/78ef2c191aee473ba38b69864919bb5c448a17be

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl (hunk 7 above)
Comment (nicholascar) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl]: Addressed by Comment https://github.com/qudt/qudt-public-repo/pull/320/commits/78ef2c191aee473ba38b69864919bb5c448a17be

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl (hunk 6 above)
Comment (nicholascar) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl]: Addressed by Comment https://github.com/qudt/qudt-public-repo/pull/320/commits/78ef2c191aee473ba38b69864919bb5c448a17be

**Comment by nicholascar** [comment]
@dr-shorthair said "Suggest using decimal" for `qudt:conversionMultiplier` x 3:

I don't want to change from the 1.0e0 as implemented as this is more inline with current QUDT examples than 1.0 would be.

There appears to be inconsistency with exponent notation use, e.g. for `unit:BTU_IT-PER-FT3` we have capital E used as such:

`unit:BTU_IT-PER-FT3 qudt:conversionMultiplier 3.725894579E04 ;`

For `unit:BTU_IT-PER-LB_F-DEG_F` we have lower case e used as such:

`unit:BTU_IT-PER-LB_F-DEG_F qudt:conversionMultiplier 4.1868e+03 ;`

I think it's out of scope for this PR to either break convention or change all notation in the vocab, so I leave it to another PR (from someone else, unless you ask nicely).

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl (hunk 8 above)
Comment (nicholascar) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:8784]: Not resolved, see comment below

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl (hunk 2 above)
Comment (nicholascar) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:18535]: Not resolved, see comment below

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl (hunk 4 above)
Comment (nicholascar) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:18547]: Not resolved, see comment below

**Comment by dr-shorthair** [comment]
There is a mixture of number types in the current catalogue. There is no consistency, so we may as well use best practice for any additions. 

And apart from the valid issues raised by @jmkeil, a TTL file is a text representation after all, so IMHO has no business telling a processor to use some (inconsistent) number of bits for the representation!

**Comment by nicholascar** [comment]
> There is a mixture of number types in the current catalogue...

OK, I have changes the 4 unit additions I've made to use decimals for their `qudt:conversionMultiplier` properties.

**Comment by dr-shorthair** [comment]
(To be quite honest, I can't see any compelling use-cases for `xsd:double` and `xsd:float`. Double and float mean different things in different systems, and XML (and TTL/N3 and JSON) are text representations after all.)

**Comment by nicholascar** [comment]
Not sure I follow: by using the un-quoted form of `1.0` & `0.0` that I have, I've encoded in the Turtle that these are decimals, so the equivalent to using the explicit `"1.0"^^xsd:decimal` etc. Since Turtle data is declarative, each RDF store etc. will (should) follow this instruction to encode the value in the way indicated - with a decimal datatype (or its underlying equivalent), so the type matters and they are not just "text representations after all".

**Comment by jhodgesatmb** [comment]
Wasn’t this bug report about the type of conversion multipliers? How did susceptibility come into it? I don’t recognize it at all, and wonder why it is all caps. Anyway, let’s talk about it.

Jack

**Comment by nicholascar** [comment]
> why it is all caps

I was following what I thought was precedence, e.g. `unit:OERSTED` & `unit:PERMITTIVITY_REL` but I'm not attached to caps at all so would you prefer `MagneticSusceptibility` & `ElectricSusceptibility`?

**Comment by jhodgesatmb** [comment]
No, you are right, that (all caps) is the qname convention. Sorry.

Jack Hodges, Ph.D.
Arbor Studios

**Comment by steveraysteveray** [comment]
OK, so to summarize where we are at this point, here are my observations (and my positions):
1. I'm convinced by @nicholascar in stating conversionMultiplier and conversionOffset values as unquoted decimals, as a valid alternative to the explicitly quoted xsd:decimal declaration.
2. I'm willing to accept the two units for magnetic and electric susceptibility, with the caveat that they point to the respective magnetic and electric susceptibility quantitykinds rather than dimensionless.
3. I agree we need a separate Issue to consolidate up our various representations of numbers toward xsd:decimal or unquoted decimals.

Sound OK?

**Comment by jhodgesatmb** [comment]
Yes

**Comment by dr-shorthair** [comment]
> I agree we need a separate Issue to consolidate up our various representations of numbers toward xsd:decimal or unquoted decimals.

It is already there - #307 

Note that ['unquoted decimal' is merely syntactic-sugar for `"NN.N"^^xsd:decimal`](https://www.w3.org/TR/turtle/#abbrev) - i.e. to an RDF parser there is no difference. 

To clarify my [grouchy comments](https://github.com/qudt/qudt-public-repo/pull/320#issuecomment-749290196) about float and double: The main point is that  XML/JSON/TTL/N3 is a textual transfer format, so `xsd:decimal` is 'honest' and also it allows for arbitrary precision. The comments by @jmkiel on `0.1` are also very relevant - if N*0.000001*1000000 does not reliably = N then we have a problem! 

In general, telling the processor what internal representation to use after parsing is intrusive and unpredictable, since it can produce different results in different architectures (though I note that `xsd:double` is defined as IEEE 64-bit, so maybe that objection is stale).

**Comment by dr-shorthair** [comment]
Fix now submitted as PR https://github.com/qudt/qudt-public-repo/pull/311

**Comment by nicholascar** [comment]
> In general, telling the processor what internal representation to use after parsing is intrusive and unpredictable

Disagree: you - the data definer - _should_ set the expectation for the literal. The "internal representation" isn't `double` `float` etc., that is an external representation. The _internal_ representation is however the system implements float, double etc. For example, in Python, there is no difference between a `float` or a `double` [1] so a Python RDF system, e.g. RDFlib, would retain `xsd:float` or `xsd:double` in the RDF literal type it stores ("external") but the "internal" representaion - a a 64 signed byte - would be the same for each.


[1] https://stackoverflow.com/questions/33960828/python-float-or-double

**Comment by nicholascar** [comment]
I support having everything in decimal though!

...and it looks like this PR is ready to go?

**Review by steveraysteveray — APPROVED** [review]

**Pull request merged**